File: app/src/main/java/vergecurrency/vergewallet/view/ui/activity/firstlaunch/PaperkeyVerificationActivity.java
package vergecurrency.vergewallet.view.ui.activity.firstlaunch;

import android.content.Intent;
import android.os.Bundle;
import android.util.Pair;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProviders;

import java.util.Arrays;

import vergecurrency.vergewallet.view.base.BaseActivity;
import vergecurrency.vergewallet.Constants;
import vergecurrency.vergewallet.R;
import vergecurrency.vergewallet.helpers.utils.MathUtils;
import vergecurrency.vergewallet.viewmodel.PaperkeyVerificationViewModel;

public class PaperkeyVerificationActivity extends BaseActivity {

	TextView firstWordCaption;
	TextView secondWordCaption;
	EditText firstWordInput;
	EditText secondWordInput;
	Button confirmButton;
	Pair<String[], int[]> verificationWords;
	String[] seed;

	PaperkeyVerificationViewModel mViewModel;

	@Override
	protected void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_paperkey_confirm_seed);

		mViewModel = ViewModelProviders.of(this).get(PaperkeyVerificationViewModel.class);

		//Get the shared preferences

		seed = mViewModel.getSeed();

		verificationWords = getTwoRandomWordsFromSeed();


		initComponents();

	}

	private void initComponents() {

		//TODO : Move this shit to viewmodel
		//set the caption for the first word according to its position
		firstWordCaption = findViewById(R.id.label_first_word);
		firstWordCaption.setText(String.format("Word #%d", verificationWords.second[0] + 1));
		//set the caption for the second word according to its position
		secondWordCaption = findViewById(R.id.label_second_word);
		secondWordCaption.setText(String.format("Word #%d", verificationWords.second[1] + 1));

		firstWordInput = findViewById(R.id.seed_first_word);
		secondWordInput = findViewById(R.id.seed_second_word);


		//assign a click listener to the
		confirmButton = findViewById(R.id.paper_key_confirm_next);
		confirmButton.setOnClickListener(onNextClick());
	}

	private Button.OnClickListener onNextClick() {
		return v -> {
			if (firstWordInput.getText().toString().equals(verificationWords.first[0].toLowerCase())) {
				if (secondWordInput.getText().toString().equals(verificationWords.first[1].toLowerCase())) {
					//Announce that it's not the first launch anymore
					//Get to the main activity
					startActivity(new Intent(getApplicationContext(), PassphraseCreationActivity.class));


				} else {
					secondWordInput.setBackgroundResource(R.color.material_red_200);
				}
			} else {
				firstWordInput.setBackgroundResource(R.color.material_red_200);
			}


		};
	}

	//TODO : Def move this shit to viewmodel
	private Pair<String[], int[]> getTwoRandomWordsFromSeed() {

		//prepare the return objects
		String[] words = new String[2];
		int[] positions = new int[2];

		//get the first position user has to check
		positions[0] = MathUtils.getRandomNumber(Constants.SEED_SIZE);
		//prepare the second one
		int second = MathUtils.getRandomNumber(Constants.SEED_SIZE);
		//avoid having twice the same position
		while (positions[0] == second) {
			second = MathUtils.getRandomNumber(Constants.SEED_SIZE);
		}
		//set the second position
		positions[1] = second;
		//sort the array
		Arrays.sort(positions);
		//assign the words to the positions
		words[0] = seed[positions[0]];
		words[1] = seed[positions[1]];
		//you don't need me to comment this.
		return new Pair<>(words, positions);
	}
}

File: app/src/main/kotlin/vergecurrency/vergewallet/view/ui/activity/firstlaunch/PaperkeyVerificationActivity.kt
package vergecurrency.vergewallet.view.ui.activity.firstlaunch

import android.content.Intent
import android.os.Bundle
import android.util.Pair
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import androidx.lifecycle.ViewModelProvider
import vergecurrency.vergewallet.Constants
import vergecurrency.vergewallet.R
import vergecurrency.vergewallet.helpers.utils.MathUtils
import vergecurrency.vergewallet.model.WalletConfiguration
import vergecurrency.vergewallet.view.base.BaseActivity
import vergecurrency.vergewallet.viewmodel.PaperkeyVerificationViewModel
import vergecurrency.vergewallet.viewmodel.WalletConfigurationFactory
import java.util.*

class PaperkeyVerificationActivity : BaseActivity() {

    private lateinit var firstWordCaption: TextView
    private lateinit var secondWordCaption: TextView
    private lateinit var firstWordInput: EditText
    private lateinit var secondWordInput: EditText
    private lateinit var confirmButton: Button
    private lateinit var verificationWords: Pair<Array<ByteArray>, IntArray>

    internal lateinit var mViewModel: WalletConfiguration

    //TODO : Def move this shit to viewmodel
    private val twoRandomWordsFromSeed: Pair<Array<ByteArray>, IntArray>
        get() {
            val words: Array<ByteArray> = Array(2) { ByteArray(0) }
            val positions = IntArray(2)
            positions[0] = MathUtils.getRandomNumber(Constants.SEED_SIZE)
            var second = MathUtils.getRandomNumber(Constants.SEED_SIZE)
            while (positions[0] == second) {
                second = MathUtils.getRandomNumber(Constants.SEED_SIZE)
            }
            positions[1] = second
            Arrays.sort(positions)
            words[0] = mViewModel.decrypt(mViewModel.getSeed()[positions[0]])
            words[1] = mViewModel.decrypt(mViewModel.getSeed()[positions[1]])
            return Pair(words, positions)
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_paperkey_confirm_seed)

        mViewModel = ViewModelProvider(this, WalletConfigurationFactory()).get(WalletConfiguration::class.java)

        //Get the shared preferences

        verificationWords = twoRandomWordsFromSeed

        initComponents()

    }

    private fun initComponents() {

        //set the caption for the first word according to its position
        firstWordCaption = findViewById(R.id.label_first_word)
        firstWordCaption.text = String.format("Word #%d", verificationWords.second[0] + 1)
        //set the caption for the second word according to its position
        secondWordCaption = findViewById(R.id.label_second_word)
        secondWordCaption.text = String.format("Word #%d", verificationWords.second[1] + 1)

        firstWordInput = findViewById(R.id.seed_first_word)
        secondWordInput = findViewById(R.id.seed_second_word)


        //assign a click listener to the
        confirmButton = findViewById(R.id.paper_key_confirm_next)
        confirmButton.setOnClickListener(onNextClick())
    }

    private fun onNextClick(): View.OnClickListener {
        return View.OnClickListener {

            if (firstWordInput.text.toString() == String(verificationWords.first[0]).toLowerCase()) {
                if (secondWordInput.text.toString() == String(verificationWords.first[1]).toLowerCase()) {

                    startActivity(Intent(applicationContext, PassphraseCreationActivity::class.java))


                } else {
                    secondWordInput.setBackgroundResource(R.color.material_red_200)
                }
            } else {
                firstWordInput.setBackgroundResource(R.color.material_red_200)
            }


        }
    }
}
